File: utils/github_client.py
import requests
import json
import os
from typing import List, Dict, Optional
import time

class GitHubClient:
    """GitHub API client for fetching repositories and SBOMs"""
    
    def __init__(self):
        self.base_url = "https://api.github.com"
        self.sbom_api_url = f"{self.base_url}/repos/{{owner}}/{{repo}}/dependency-graph/sbom"
        self.github_token = os.environ.get('GITHUB_TOKEN')
        self.headers = {}
        
        # Set up headers with authentication if token is available
        if self.github_token:
            self.headers = {
                'Authorization': f'token {self.github_token}',
                'Accept': 'application/vnd.github.v3+json'
            }
        else:
            self.headers = {
                'Accept': 'application/vnd.github.v3+json'
            }
            print("⚠️  No GitHub token provided. Using unauthenticated requests (rate limited to 60 requests/hour)")
    
    def _handle_rate_limit(self, response):
        """Handle GitHub API rate limiting with persistent retry"""
        if response.status_code == 403:
            # Check if it's a rate limit error
            rate_limit_remaining = response.headers.get('X-RateLimit-Remaining')
            rate_limit_reset = response.headers.get('X-RateLimit-Reset')
            
            if rate_limit_remaining == '0':
                if rate_limit_reset:
                    reset_time = int(rate_limit_reset)
                    wait_time = reset_time - int(time.time())
                    if wait_time > 0:
                        print(f"⏳ Rate limit exceeded. Waiting {wait_time} seconds for reset...")
                        time.sleep(wait_time + 2)  # Add 2 second buffer
                        print("✅ Rate limit reset. Continuing...")
                        return True  # Retry the request
                    else:
                        print("⚠️  Rate limit exceeded. Please wait before trying again.")
                        return False
                else:
                    print("⚠️  Rate limit exceeded. Please wait before trying again.")
                    return False
            else:
                if self.github_token:
                    print("⚠️  API access denied. Your token may not have sufficient permissions.")
                else:
                    print("⚠️  API access denied. This organization may require authentication.")
                return False
        
        return False
    
    def _check_rate_limit_remaining(self, response):
        """Check if we're running low on rate limit"""
        rate_limit_remaining = response.headers.get('X-RateLimit-Remaining')
        if rate_limit_remaining and int(rate_limit_remaining) <= 5:
            print(f"⚠️  Rate limit running low: {rate_limit_remaining} requests remaining")
            return True
        return False
    
    def _make_request(self, url: str, retry_count: int = 0) -> Optional[requests.Response]:
        """Make a request with rate limit handling"""
        if retry_count > 3:
            print("❌ Too many retries. Stopping request.")
            return None
            
        try:
            response = requests.get(url, headers=self.headers)
            
            # Handle rate limiting
            if self._handle_rate_limit(response):
                if retry_count < 3:
                    return self._make_request(url, retry_count + 1)
                else:
                    return None
            
            return response
            
        except requests.exceptions.RequestException as e:
            print(f"❌ Request failed: {e}")
            return None
    
    def get_repositories(self, org_name: str) -> List[Dict]:
        """Fetch all public repositories for a GitHub organization"""
        url = f"{self.base_url}/orgs/{org_name}/repos"
        response = self._make_request(url)
        
        if not response:
            if self.github_token:
                raise Exception("Failed to fetch repositories. Please check your GitHub token and try again.")
            else:
                raise Exception("Failed to fetch repositories. Consider setting a GitHub token for higher rate limits.")
        
        if response.status_code != 200:
            if response.status_code == 404:
                raise Exception(f"Organization '{org_name}' not found. Please check the organization name.")
            elif response.status_code == 403:
                if self.github_token:
                    raise Exception("Access denied. The organization might be private or your token may not have sufficient permissions.")
                else:
                    raise Exception("Access denied. This organization may require authentication. Please set a GitHub token.")
            else:
                raise Exception(f"Failed to fetch repositories: {response.status_code} {response.text}")
        
        return response.json()
    
    def fetch_sbom(self, owner: str, repo: str, output_dir: str = "sboms") -> Optional[str]:
        """Download SBOM for a specific repository"""
        url = self.sbom_api_url.format(owner=owner, repo=repo)
        response = self._make_request(url)

        if not response:
            print(f"⚠️  Could not fetch SBOM for {owner}/{repo} due to rate limiting or network issues")
            return None

        # Check if we're running low on rate limit
        if response.status_code == 200:
            self._check_rate_limit_remaining(response)
            sbom_data = response.json()
            
            # Ensure output directory exists
            if not os.path.exists(output_dir):
                os.makedirs(output_dir)
            
            # Save SBOM to file
            sbom_filename = os.path.join(output_dir, f"{owner}_{repo}_sbom.json")
            with open(sbom_filename, 'w') as f:
                json.dump(sbom_data, f, indent=2)
            
            return sbom_filename
        elif response.status_code == 404:
            print(f"ℹ️  SBOM not available for {owner}/{repo} (dependency graph not enabled)")
            return None
        elif response.status_code == 403:
            print(f"⚠️  Access denied for {owner}/{repo} (private repository or insufficient permissions)")
            return None
        elif response.status_code == 401:
            print(f"⚠️  Authentication required for {owner}/{repo} (private repository)")
            return None
        elif response.status_code == 422:
            print(f"ℹ️  SBOM not available for {owner}/{repo} (dependency graph not enabled)")
            return None
        else:
            print(f"⚠️  SBOM not available for {owner}/{repo}: {response.status_code} {response.text}")
            return None
    
    def get_public_repositories(self, org_name: str) -> List[Dict]:
        """Get only public repositories for an organization"""
        repos = self.get_repositories(org_name)
        return [repo for repo in repos if repo['visibility'] == 'public']
    
    def get_repository_count(self, org_name: str) -> int:
        """Get the count of public repositories for an organization"""
        return len(self.get_public_repositories(org_name))
    
    def get_rate_limit_info(self) -> Dict[str, str]:
        """Get current rate limit information"""
        url = f"{self.base_url}/rate_limit"
        response = self._make_request(url)
        
        if response and response.status_code == 200:
            data = response.json()
            core = data.get('resources', {}).get('core', {})
            return {
                'limit': str(core.get('limit', 'Unknown')),
                'remaining': str(core.get('remaining', 'Unknown')),
                'reset': str(core.get('reset', 'Unknown')),
                'authenticated': 'Yes' if self.github_token else 'No'
            }
        else:
            return {
                'limit': 'Unknown',
                'remaining': 'Unknown',
                'reset': 'Unknown',
                'authenticated': 'No'
            } 

File: docs/js/github-client.js
/**
 * GitHub Client - Handles GitHub API requests with rate limiting
 */
class GitHubClient {
    constructor() {
        this.baseUrl = 'https://api.github.com';
        this.token = null; // Don't load from localStorage
        this.headers = {
            'Accept': 'application/vnd.github.v3+json'
        };
        
        // Custom event emitter for rate limit events
        this.eventTarget = new EventTarget();
    }

    /**
     * Add event listener for rate limit events
     */
    addEventListener(type, listener) {
        this.eventTarget.addEventListener(type, listener);
    }

    /**
     * Remove event listener
     */
    removeEventListener(type, listener) {
        this.eventTarget.removeEventListener(type, listener);
    }

    /**
     * Dispatch custom event
     */
    dispatchEvent(event) {
        this.eventTarget.dispatchEvent(event);
    }

    /**
     * Set GitHub token (not persisted)
     */
    setToken(token) {
        this.token = token;
        if (token) {
            this.headers['Authorization'] = `token ${token}`;
        } else {
            delete this.headers['Authorization'];
        }
        // Don't save to localStorage
    }

    /**
     * Get repositories for an organization or user
     */
    async getRepositories(ownerName) {
        // Try organization endpoint first
        let url = `${this.baseUrl}/orgs/${ownerName}/repos?per_page=100`;
        let response = await this.makeRequest(url);
        
        if (response.ok) {
            console.log(`✅ Found organization: ${ownerName}`);
            const repos = await this.getAllPages(url);
            return repos.filter(repo => repo.visibility === 'public');
        }
        
        // If organization fails, try user endpoint
        if (response.status === 404) {
            console.log(`ℹ️  Not found as organization, trying as user: ${ownerName}`);
            url = `${this.baseUrl}/users/${ownerName}/repos?per_page=100`;
            response = await this.makeRequest(url);
            
            if (response.ok) {
                console.log(`✅ Found user: ${ownerName}`);
                const repos = await this.getAllPages(url);
                return repos.filter(repo => repo.visibility === 'public');
            }
        }
        
        // If both fail, throw appropriate error
        if (response.status === 404) {
            throw new Error(`Organization or user '${ownerName}' not found`);
        } else if (response.status === 403) {
            throw new Error('Access denied. The organization/user might be private or require authentication.');
        } else {
            throw new Error(`Failed to fetch repositories: ${response.status} ${response.statusText}`);
        }
    }

    /**
     * Get all pages from a paginated API endpoint
     */
    async getAllPages(url) {
        const allRepos = [];
        let currentUrl = url;
        
        while (currentUrl) {
            console.log(`📄 Fetching page: ${currentUrl}`);
            const response = await this.makeRequest(currentUrl);
            
            if (!response.ok) {
                throw new Error(`Failed to fetch repositories: ${response.status} ${response.statusText}`);
            }
            
            const repos = await response.json();
            allRepos.push(...repos);
            
            // Check for next page
            const linkHeader = response.headers.get('Link');
            if (linkHeader) {
                const nextMatch = linkHeader.match(/<([^>]+)>;\s*rel="next"/);
                if (nextMatch) {
                    currentUrl = nextMatch[1];
                } else {
                    currentUrl = null; // No next page
                }
            } else {
                currentUrl = null; // No Link header, assume single page
            }
        }
        
        console.log(`📊 Total repositories fetched: ${allRepos.length}`);
        return allRepos;
    }

    /**
     * Fetch SBOM for a repository
     */
    async fetchSBOM(owner, repo) {
        const url = `${this.baseUrl}/repos/${owner}/${repo}/dependency-graph/sbom`;
        const response = await this.makeRequest(url);
        
        if (!response.ok) {
            if (response.status === 404) {
                console.log(`ℹ️  SBOM not available for ${owner}/${repo} (dependency graph not enabled or repository not found)`);
                return null;
            } else if (response.status === 403) {
                console.log(`⚠️  Access denied for ${owner}/${repo} (private repository or insufficient permissions)`);
                return null;
            } else if (response.status === 401) {
                console.log(`⚠️  Authentication required for ${owner}/${repo} (private repository)`);
                return null;
            } else if (response.status === 422) {
                console.log(`ℹ️  SBOM not available for ${owner}/${repo} (dependency graph not enabled)`);
                return null;
            } else if (response.status === 429) {
                console.log(`⏳ Rate limit exceeded for ${owner}/${repo}, will retry later`);
                return null;
            } else {
                console.log(`⚠️  SBOM not available for ${owner}/${repo}: ${response.status} ${response.statusText}`);
                return null;
            }
        }
        
        const sbomData = await response.json();
        
        // Validate SBOM data structure
        if (!sbomData || !sbomData.sbom || !sbomData.sbom.packages) {
            console.log(`ℹ️  SBOM data is empty for ${owner}/${repo}`);
            return null;
        }
        
        // Log success with dependency count
        const packageCount = sbomData.sbom.packages.length;
        console.log(`✅ SBOM fetched for ${owner}/${repo}: ${packageCount} packages found`);
        
        return sbomData;
    }

    /**
     * Get rate limit information
     */
    async getRateLimitInfo() {
        const url = `${this.baseUrl}/rate_limit`;
        const response = await this.makeRequest(url);
        
        if (response.ok) {
            const data = await response.json();
            const core = data.resources.core;
            return {
                limit: core.limit,
                remaining: core.remaining,
                reset: core.reset,
                authenticated: this.token ? 'Yes' : 'No'
            };
        } else {
            return {
                limit: 'Unknown',
                remaining: 'Unknown',
                reset: 'Unknown',
                authenticated: 'No'
            };
        }
    }

    /**
     * Save rate limit state to localStorage
     */
    saveRateLimitState(waitTime, resetTime) {
        const state = {
            waitTime,
            resetTime,
            timestamp: Date.now(),
            organization: localStorage.getItem('current_analysis_org') || 'unknown'
        };
        localStorage.setItem('rate_limit_state', JSON.stringify(state));
    }

    /**
     * Load rate limit state from localStorage
     */
    loadRateLimitState() {
        const state = localStorage.getItem('rate_limit_state');
        return state ? JSON.parse(state) : null;
    }

    /**
     * Clear rate limit state
     */
    clearRateLimitState() {
        localStorage.removeItem('rate_limit_state');
    }

    /**
     * Make a request with rate limit handling
     */
    async makeRequest(url, retryCount = 0) {
        if (retryCount > 3) {
            console.log('❌ Too many retries. Stopping request.');
            return new Response(null, { status: 429 });
        }
        
        try {
            const response = await fetch(url, {
                method: 'GET',
                headers: this.headers
            });
            
            // Handle rate limiting
            if (response.status === 403) {
                const remaining = response.headers.get('X-RateLimit-Remaining');
                const reset = response.headers.get('X-RateLimit-Reset');
                
                if (remaining === '0' && reset) {
                    const resetTime = parseInt(reset);
                    const waitTime = resetTime - Math.floor(Date.now() / 1000);
                    
                    if (waitTime > 0) {
                        // Save rate limit state
                        this.saveRateLimitState(waitTime, resetTime);
                        
                        // Dispatch rate limit event
                        this.dispatchEvent(new CustomEvent('rateLimitExceeded', {
                            detail: {
                                waitTime,
                                resetTime,
                                resetDate: new Date(resetTime * 1000)
                            }
                        }));
                        
                        console.log(`⏳ Rate limit exceeded. Waiting ${waitTime} seconds for reset...`);
                        await this.sleep(waitTime + 2); // Add 2 second buffer
                        
                        // Clear rate limit state after waiting
                        this.clearRateLimitState();
                        
                        // Dispatch rate limit reset event
                        this.dispatchEvent(new CustomEvent('rateLimitReset'));
                        
                        console.log('✅ Rate limit reset. Continuing...');
                        return this.makeRequest(url, retryCount + 1);
                    }
                }
            }
            
            return response;
        } catch (error) {
            console.log(`❌ Request failed: ${error.message}`);
            return new Response(null, { status: 500 });
        }
    }

    /**
     * Sleep utility function
     */
    sleep(seconds) {
        return new Promise(resolve => setTimeout(resolve, seconds * 1000));
    }

    /**
     * Check if rate limit is running low
     */
    async checkRateLimitRemaining(response) {
        const remaining = response.headers.get('X-RateLimit-Remaining');
        if (remaining && parseInt(remaining) <= 5) {
            console.log(`⚠️  Rate limit running low: ${remaining} requests remaining`);
            return true;
        }
        return false;
    }
}

// Export for use in other modules
window.GitHubClient = GitHubClient; 